File: proxy-service/app/schemas/proxy.py
from pydantic import BaseModel, Field
from typing import Optional, List, Generic, TypeVar
from datetime import datetime

T = TypeVar('T')

class PaginatedResponse(BaseModel, Generic[T]):
    """通用分页响应模式"""
    items: List[T]
    total: int
    page: int
    size: int
    pages: int

    @classmethod
    def create(cls, items: List[T], total: int, page: int, size: int):
        """创建分页响应"""
        pages = (total + size - 1) // size if total > 0 else 0
        return cls(
            items=items,
            total=total,
            page=page,
            size=size,
            pages=pages
        )


class ProxyBase(BaseModel):
    type: str
    ip: str
    port: int
    username: Optional[str] = None
    password: Optional[str] = None
    country: Optional[str] = None
    city: Optional[str] = None


class ProxyCreate(ProxyBase):
    pass


class ProxyUpdate(BaseModel):
    type: Optional[str] = None
    ip: Optional[str] = None
    port: Optional[int] = None
    username: Optional[str] = None
    password: Optional[str] = None
    country: Optional[str] = None
    city: Optional[str] = None
    isp: Optional[str] = None
    latency: Optional[int] = None
    success_rate: Optional[float] = None
    status: Optional[str] = None


class ProxyQualityBase(BaseModel):
    total_usage: int
    success_count: int
    quality_score: float
    last_used: Optional[datetime] = None
    cooldown_time: int


class ProxyQualityCreate(BaseModel):
    proxy_id: str
    total_usage: int = 0
    success_count: int = 0
    quality_score: float = 0.8
    last_used: Optional[datetime] = None
    cooldown_time: int = 0


class ProxyQualityUpdate(BaseModel):
    total_usage: Optional[int] = None
    success_count: Optional[int] = None
    quality_score: Optional[float] = None
    last_used: Optional[datetime] = None
    cooldown_time: Optional[int] = None


class ProxyQualityResponse(ProxyQualityBase):
    id: str
    proxy_id: str
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ProxyResponse(ProxyBase):
    id: str
    isp: Optional[str] = None
    latency: Optional[int] = None
    success_rate: float
    last_check: Optional[datetime] = None
    status: str
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ProxyWithQualityResponse(ProxyResponse):
    """包含质量信息的代理响应"""
    quality: Optional[ProxyQualityResponse] = None

    class Config:
        from_attributes = True


class ProxyCheck(BaseModel):
    proxy_ids: List[str]


class ProxyCheckResult(BaseModel):
    id: str
    status: str
    latency: Optional[int] = None
    success_rate: Optional[float] = None
    error_msg: Optional[str] = None


class ProxyCheckResponse(BaseModel):
    total: int
    checked: int
    active: int
    inactive: int
    results: List[ProxyCheckResult]


class ProxyImport(BaseModel):
    proxies: List[ProxyCreate]
    check_availability: bool = True


class ProxyImportResult(BaseModel):
    id: str
    type: str
    ip: str
    status: str


class ProxyImportResponse(BaseModel):
    total: int
    imported: int
    active: int
    inactive: int
    proxies: List[ProxyImportResult]


class ProxyUsageResult(BaseModel):
    success: bool
    error_msg: Optional[str] = None


# 代理使用历史记录相关的schema
class ProxyUsageHistoryCreate(BaseModel):
    proxy_id: str
    task_id: Optional[str] = None
    account_id: Optional[str] = None
    service_name: Optional[str] = None
    success: str = "SUCCESS"  # SUCCESS, FAILED, TIMEOUT
    response_time: Optional[int] = None
    
    proxy_ip: Optional[str] = None
    proxy_port: Optional[int] = None
    account_username_email: Optional[str] = None
    task_name: Optional[str] = None
    quality_score: Optional[float] = None
    latency: Optional[int] = None


class ProxyUsageHistoryResponse(BaseModel):
    id: str
    proxy_id: str
    # 原有字段
    account_id: Optional[str] = None
    task_id: Optional[str] = None
    service_name: Optional[str] = None
    success: str
    response_time: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    # 冗余字段（直接存储，无需关联查询）
    proxy_ip: Optional[str] = None
    proxy_port: Optional[int] = None
    account_username_email: Optional[str] = None
    task_name: Optional[str] = None
    quality_score: Optional[float] = None
    latency: Optional[int] = None

    class Config:
        from_attributes = True


class ProxyUsageHistoryFilter(BaseModel):
    proxy_id: Optional[str] = None
    proxy_ip: Optional[str] = None
    account_email: Optional[str] = None
    task_name: Optional[str] = None
    service_name: Optional[str] = None
    success: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    page: int = 1
    size: int = 20

# 新增分页响应类型
class ProxyListResponse(PaginatedResponse[ProxyResponse]):
    """代理列表分页响应"""
    pass

class ProxyQualityListResponse(PaginatedResponse[ProxyQualityResponse]):
    """代理质量列表分页响应"""
    pass


class ProxyQualityInfoResponse(BaseModel):
    """代理质量信息综合响应 - 包含IP+port、检测成功率、使用次数、成功次数、质量分数、最近使用时间"""
    # 代理基本信息
    proxy_id: str
    ip: str
    port: int
    proxy_type: str
    
    # 质量信息
    total_usage: int = Field(description="总使用次数")
    success_count: int = Field(description="成功次数")
    success_rate: float = Field(description="检测成功率")
    quality_score: float = Field(description="质量分数")
    last_used: Optional[datetime] = Field(description="最近使用时间")
    
    # 其他信息
    status: str = Field(description="代理状态")
    country: Optional[str] = Field(description="国家")
    city: Optional[str] = Field(description="城市")
    isp: Optional[str] = Field(description="ISP")
    latency: Optional[int] = Field(description="延迟(毫秒)")
    created_at: datetime = Field(description="创建时间")
    updated_at: Optional[datetime] = Field(description="更新时间")

    class Config:
        from_attributes = True


class ProxyQualityInfoListResponse(PaginatedResponse[ProxyQualityInfoResponse]):
    """代理质量信息列表分页响应"""
    pass

class ProxyUsageHistoryListResponse(PaginatedResponse[ProxyUsageHistoryResponse]):
    """代理使用历史记录列表分页响应"""
    pass

File: frontend-service/src/types/proxy.ts
import type { BaseEntity, PaginationParams, PaginatedResponse } from './common'

// 代理类型 - 注意API返回的是大写
export type ProxyType = 'HTTP' | 'HTTPS' | 'SOCKS4' | 'SOCKS5'

// 代理状态
export type ProxyStatus = 'ACTIVE' | 'INACTIVE'

// 代理实体 - 匹配真实API响应格式
export interface Proxy {
  id: string
  type: ProxyType
  ip: string // API中使用ip而不是host
  port: number
  username?: string
  password?: string
  country?: string // 所属国家
  city?: string // 城市信息，API中有这个字段
  isp?: string | null // ISP供应商
  latency?: number // 网络延迟 (ms)
  success_rate?: number // 成功率 (0-1)
  last_check?: string // 最后检查时间
  status: ProxyStatus
  created_at: string // 创建时间
  updated_at: string // 更新时间
}

// 创建代理参数
export interface ProxyCreateParams {
  ip: string
  port: number
  type: ProxyType
  username?: string
  password?: string
  country?: string
  city?: string
  isp?: string
}

// 更新代理参数
export interface ProxyUpdateParams {
  ip?: string
  port?: number
  type?: ProxyType
  username?: string
  password?: string
  status?: ProxyStatus
  country?: string
  city?: string
  isp?: string
}

// 代理质量信息 - 匹配真实API响应格式
export interface ProxyQuality {
  proxy_id: string
  ip: string
  port: number
  proxy_type: ProxyType // 代理类型
  total_usage: number // 总使用次数
  success_count: number // 成功次数
  success_rate: number // 成功率 (0-1)
  quality_score: number // 质量分数 (0-1)
  last_used: string | null // 最近使用时间
  status: ProxyStatus // 代理状态
  country?: string // 国家
  city?: string // 城市
  isp?: string | null // ISP供应商
  latency: number // 延迟(ms)
  created_at: string // 创建时间
  updated_at: string // 更新时间
}

// 代理历史记录
export interface ProxyHistory extends BaseEntity {
  id: string // 历史记录ID
  proxy_id: string // 代理ID
  proxy_ip: string // 代理IP
  account_email?: string // 账户邮箱
  account_id: string // 账户ID
  task_name?: string // 任务名称
  task_id: string // 任务ID
  service_name: string // 服务名称
  success: string // 是否成功
  response_time: number // 响应时长 (ms)
  created_at: string // 创建时间
  updated_at: string // 更新时间
}

// 代理统计
export interface ProxyStats {
  total: number
  active: number
  inactive: number
  byType: Record<ProxyType, number>
  byCountry: Record<string, number>
  byRegion: Record<string, number>
  averageLatency: number
  totalUsageCount: number
}

// 导出代理类型别名
export type { PaginationParams, PaginatedResponse } 

export interface ProxyCheckResponse {
  total: number
  checked: number
  active: number
  inactive: number
  results: ProxyCheckResult[]
}

export interface ProxyCheckResult {
  id: string
  status: string
  latency?: number
  success_rate?: number
  error_msg?: string
}

export interface ProxyQualityListResponse {
  items: ProxyQuality[]
  total: number
  page: number
  size: number
  pages: number
}

export interface ProxyQualityInfo {
  proxy_id: string
  ip: string
  port: number
  proxy_type: ProxyType
  total_usage: number
  success_count: number
  success_rate: number
  quality_score: number
  last_used: string | null
}

export interface ProxyQualityListParams {
  page?: number
  size?: number
  status?: string
  country?: string
  min_quality_score?: string
  sort_by?: string
  sort_order?: string
}

export interface ProxyHistoryListParams {
  page?: number
  size?: number
  status?: string
  country?: string
  min_quality_score?: string
  success?: string
  service_name?: string
  start_date?: string
  end_date?: string
  account_email?: string
  task_name?: string
}

export interface ProxyHistoryListResponse {
  items: ProxyHistory[]
  total: number
  page: number
  size: number
  pages: number
}

export interface ProxyHistoryItem {
  id: string
  proxy_id: string
  proxy_ip: string
  account_email?: string
  account_id: string
  task_name?: string
  task_id: string
  service_name: string
  success: string
  response_time: number
  created_at: string
  updated_at: string
}